### PR TITLE
[KEP-2400]: Restrict access to swap for containers in high priority Pods

### PR DESCRIPTION
**Pull request by iholder101**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
Exclude critical pods from gaining swap access.

I believe this is valuable for two main reasons:
1. Critical pods are assumed to not tolerate performance derogations which could be impacted swap access.
2. This provides another way of opting-out burstable pods from swap access.

p.s. currently, it is possible to opt-out of swap for burstable pods by setting `requests.memory == limits.memory`. However, this approach forces the workload owner to set limits which is unacceptable for certain workloads. With this, an administrator can choose to classify such burstable pods as critical to opt-out of swap.

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
Changed Linux swap handling to restrict access to swap for containers in high priority Pods.
New Pods that have a node- or cluster-critical priority are prohibited from accessing swap on Linux,
even if your cluster and node configuration could otherwise allow this.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- KEP: https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2400-node-swap
```

### CODE REVIEW

**Comment by iholder101** [comment]
@haircommander do we need a backing issue for this PR?

**Comment by sftim** [comment]
Changelog suggestion
```diff
-Exclude critical pods from having swap access
+Changed Linux swap handling to restrict access to swap for containers in high priority Pods.
+New Pods that have a node- or cluster-critical priority are prohibited from accessing swap on Linux,
+even if your cluster and node configuration could otherwise allow this.
```

**Comment by iholder101** [comment]
> Changelog suggestion
> 
> ```diff
> -Exclude critical pods from having swap access
> +Changed Linux swap handling to restrict access to swap for containers in high priority Pods.
> +New Pods that have a node- or cluster-critical priority are prohibited from accessing swap on Linux,
> +even if your cluster and node configuration could otherwise allow this.
> ```

Thanks, done!

**Comment by haircommander** [comment]
> @haircommander do we need a backing issue for this PR?

I think we can discuss pros and cons here

**Comment by yujuhong** [comment]
> > @haircommander do we need a backing issue for this PR?
> 
> I think we can discuss pros and cons here

I think the use case is reasonable. Many critical components don't have the memory limit set today. 
That said, I wonder if we make too many indirect/internal decisions for swap instead of allowing it to surface on the API for the users to configure.

**Comment by aojea** [comment]
Is this in line with the KEP https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2400-node-swap#set-aside-swap-for-system-critical-daemons

> > Set Aside Swap for System Critical Daemons
> Note In Beta2, we found that having system critical daemons swapping memory could cause degration of services.
> 
> System critical daemons (such as Kubelet) are essential for node health. Usually, an appropriate portion of system resources (e.g., memory, CPU) is reserved as system reserved. However, swap doesn't inherently support reserving a portion out of the total available. For instance, in the case of memory, we set memory.min on the node-level cgroup to ensure an adequate amount of memory is set aside, away from the pods, and for system critical daemons. But there is no equivalent for swap; i.e., no memory.swap.min is supported in the kernel.
> 
> Since this proposal advocates enabling swap only for the Burstable QoS pods, this can be done by setting memory.swap.max on the cgroups used by the Burstable QoS pods. The value of this memory.swap.max can be calculated by:
> 
> memory.swap.max = total swap memory available on the system - system reserve (memory)
> 
> This is the total amount of swap available for all the Burstable QoS pods; let's call it TotalPodsSwapAvailable. This will ensure that the system critical daemons will have access to the swap at least equal to the system reserved memory. This will indirectly act as having support for swap in system reserved.

**Comment by fabiand** [comment]
@yujuhong PRs like this will help the system to make reasonable - or even better: good - decisions about how to leverage swap _in the absence of further informations_.

Thus to me this PR is orthorgonal to any API that can/could be introduced to manage swap on a workload basis.

**Comment by yujuhong** [comment]
> > I agree with @yujuhong point. I think forbidding critical pods to swap is valid but I think that other companies/users may want to forbid other types of pods to swap.
> > I think defining a pod api to forbid swap and we could set that field for system critical pods.
> > This current implementation is too implicit and I think we should have a declarative approach for pods to say that swap is disabled so other users can build more complicated rules than this.
> 
> Thanks for raising this topic.
> 
> Adding APIs / customizability for swap is a big and complicated topic, and it's not the only way of further extending the swap feature.
> 
> This issue was raised today at the sig-node meeting. There I've suggested to frame the current KEP as "basic swap enablement" which will later be further extended. Important issues like customizability will be addressed in follow-up KEPs that will be dedicated to these topics, which I'm already working on. In the meeting there was a broad consensus that this approach makes sense.
> 
> I am currently working on a PR to update the KEP which will be ready soon.

I think at the least, we need to make sure this basic swap capability will be compatible with the future extension. It might be worth updating the KEP to reflect that :) 

On the other hand, I think the lack of API fields to configure and surface swap-related functions and status remain a larger issue. Just saw this question from Tim in another PR: https://github.com/kubernetes/kubernetes/pull/125278#issuecomment-2148303689

**Comment by iholder101** [comment]
> > > I agree with @yujuhong point. I think forbidding critical pods to swap is valid but I think that other companies/users may want to forbid other types of pods to swap.
> > > I think defining a pod api to forbid swap and we could set that field for system critical pods.
> > > This current implementation is too implicit and I think we should have a declarative approach for pods to say that swap is disabled so other users can build more complicated rules than this.
> > 
> > 
> > Thanks for raising this topic.
> > Adding APIs / customizability for swap is a big and complicated topic, and it's not the only way of further extending the swap feature.
> > This issue was raised today at the sig-node meeting. There I've suggested to frame the current KEP as "basic swap enablement" which will later be further extended. Important issues like customizability will be addressed in follow-up KEPs that will be dedicated to these topics, which I'm already working on. In the meeting there was a broad consensus that this approach makes sense.
> > I am currently working on a PR to update the KEP which will be ready soon.
> 

> I think at the least, we need to make sure this basic swap capability will be compatible with the future extension.
> It might be worth updating the KEP to reflect that :)

Absolutely. I'm working on a KEP update. :)

My guiding principle is that until we introduce APIs for customizations the behavior should be kept as before.
For example, in this PR, the high-priority pods won't have swap access (which is the same behavior as without using NodeSwap anyways). This way we open the door for a follow-up thorough discussion that's focused on APIs and customization. I believe this approach is best to ensure being compatible with further extentions in the near future.

> On the other hand, I think the lack of API fields to configure and surface swap-related functions and status remain a larger issue. Just saw this question from Tim in another PR: [#125278 (comment)](https://github.com/kubernetes/kubernetes/pull/125278#issuecomment-2148303689)

I believe this is related to a different and specific topic.

Long story short:
* I've created and merged https://github.com/kubernetes/kubernetes/pull/118865 for adding swap metrics and summary APIs.
* While updating the KEP, we [received feedback](https://github.com/kubernetes/enhancements/pull/4401#discussion_r1479124963) from Dawn to add swap to `kubectl describe node`'s output.
* I've implemented it here https://github.com/kubernetes/kubernetes/pull/123963, but then received push-back from Jordan Liggit. Finally, Dawn and Jordan agreed that adding a condition is the way to go.
* Now Tim's pushing back on the condition and suggests to add a node field instead.

TLDR: Essentially, there are several alternatives to provide swap-related information via `kubectl describe node` and the sig-node leads are discussing which is best. But this is a different subject than customizability or APIs.

**Comment by iholder101** [comment]
> > I agree with @yujuhong point. I think forbidding critical pods to swap is valid but I think that other companies/users may want to forbid other types of pods to swap.
> > I think defining a pod api to forbid swap and we could set that field for system critical pods.
> > This current implementation is too implicit and I think we should have a declarative approach for pods to say that swap is disabled so other users can build more complicated rules than this.
> 
> Thanks for raising this topic.
> 
> Adding APIs / customizability for swap is a big and complicated topic, and it's not the only way of further extending the swap feature.
> 
> This issue was raised today at the sig-node meeting. There I've suggested to frame the current KEP as "basic swap enablement" which will later be further extended. Important issues like customizability will be addressed in follow-up KEPs that will be dedicated to these topics, which I'm already working on. In the meeting there was a broad consensus that this approach makes sense.
> 
> I am currently working on a PR to update the KEP which will be ready soon.

@yujuhong @mrunalp @SergeyKanzhelev 
We reached an understanding here during the sig-node meeting that this approach makes sense.

Is there anything else missing here? would you be able to have a look?

**Comment by yujuhong** [comment]
The current change looks okay to me. 

Like mentioned in https://github.com/kubernetes/kubernetes/pull/125277#issuecomment-2147853612, if we do extend the API to customize this further, I'd expect we also surface/set system-critical pods that way to avoid confusion.

**Comment by fabiand** [comment]
> Like mentioned in https://github.com/kubernetes/kubernetes/pull/125277#issuecomment-2147853612, if we do extend the API to customize this further, I'd expect we also surface/set system-critical pods that way to avoid confusion.

If / Once swpa is exposed as an API, then I'd expect that

a. any pod can define it's exact swap usage, including critical pods.
b. if a pod does not specify it's swap usage, then the here proposed heuristic will still be applied to critical pods

**Comment by iholder101** [comment]
> The current change looks okay to me.
> 
> Like mentioned in [#125277 (comment)](https://github.com/kubernetes/kubernetes/pull/125277#issuecomment-2147853612), if we do extend the API to customize this further, I'd expect we also surface/set system-critical pods that way to avoid confusion.

Thanks @yujuhong for raising that, I fully agree and will definitely address that in the follow-up KEP (that is still WIP).

Would you mind LGTMing this PR then? Or is there anything missing?

**Comment by yujuhong** [comment]
@SergeyKanzhelev @mrunalp for test/e2e_node approval

**Review by mrunalp — APPROVED** [review]

**Comment by kannon92** [comment]
Job sounds like a legit failure.

ERROR: pkg/kubelet/kuberuntime/kuberuntime_container_linux_test.go:1262:5: bool-compare: use assert.True (testifylint)
ERROR: 				assert.Equal(t, true, types.IsCriticalPod(pod), "pod is expected to be critical")
ERROR: 				^
 Please review the above warnings. You can test via

**Comment by pacoxu** [comment]
/hold 
for ci failure

File: pkg/kubelet/kuberuntime/kuberuntime_container_linux_test.go [hunk 1]
```diff
@@ -1244,6 +1257,11 @@ func TestGenerateLinuxContainerResourcesWithSwap(t *testing.T) {
 			pod.Spec.Containers[0].Resources = resourceReqsC1
 			pod.Spec.Containers[1].Resources = resourceReqsC2
 
+			if tc.isCriticalPod {
+				pod.Spec.Priority = ptr.To(scheduling.SystemCriticalPriority)
+				assert.Equal(t, true, types.IsCriticalPod(pod), "pod is expected to be critical")
```
Comment (pacoxu) [inline pkg/kubelet/kuberuntime/kuberuntime_container_linux_test.go:1262]:
```suggestion
				assert.True(t, types.IsCriticalPod(pod), "pod is expected to be critical")
```

**Comment by iholder101** [comment]
Thank you @pacoxu and @kannon92 for the notice!
Sorry for not noticing this myself, I'm out of work for an extended period.
